### PR TITLE
[FIX] website: improve mega menus usability

### PR DESCRIPTION
**Pull request by younn-o**
This commit adds the possibility to modify some background colors of the
mega menus so that the templates are fully editable.

Some of these templates have a column with an extended background, using
the :before css pseudo element.
As this is not a DOM element, it cannot be changed from javascript.
It was decided that using shapes would not produce the same effect as
the designer intended, and that it would be too confusing for the user
to be provided with an option allowing to extend the background of a
column.
Therefore, these specific s_mega_menu_gray_area elements must have their
background hardcoded.

task-2668908

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Some general comments but did not go too deep in my review because: this work should be done for 15.0 as a stable fix and most of this can't be merged in 15.0 :wink: So check how to do a stable fix to solve the problem and potentially keep this PR for the proper master fix version if necessary :+1:

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/website/static/src/scss/website.scss [hunk 1]
```diff
@@ -1187,6 +1187,27 @@ header {
...
+            width: 100vw;
+            height: 100%;
+            position: absolute;
+            left: 0;
+            top: 0;
+            z-index: -1;
+            pointer-events: none;
+            background: rgba(0, 0, 0, .05);
+        }
+    }
```
Comment (qsm-odoo) [inline addons/website/static/src/scss/website.scss:1210]: Sorry I changed my mind about this (note that originally I asked to keep a comment in the XXX.scss file to indicate that this rule exist so it would not be forgotten if we ever do a XXX+1.scss file which does not need it anymore). I don't like "ugly" code to be standard. Instead, make a [placeholder rule](https://sass-lang.com/documentation/at-rules/extend#placeholder-selectors) and extend it in the appropriate XXX.scss templates). That way we still share the code, we still create only one rule about this but we keep it specific to each template.

At some point we should disable the scss of mega menu templates which are not used by the user (like we with header templates) so it will make even more sense to do it this way.

File: addons/website/views/snippets/s_mega_menu_odoo_menu.xml [hunk 2]
```diff
@@ -2,10 +2,10 @@
 <odoo>
 
 <template id="s_mega_menu_odoo_menu" name="Odoo Menu" groups="base.group_user">
-    <section class="s_mega_menu_odoo_menu pt16 o_colored_level o_cc o_cc1">
+    <section class="s_mega_menu_odoo_menu pt16 o_colored_level o_cc o_cc1" data-vcss="001">
```
Comment (qsm-odoo) [inline addons/website/views/snippets/s_mega_menu_odoo_menu.xml:5]: :grimacing: :fearful: Probably a mistake, never in stable anyway

File: addons/website/views/snippets/s_mega_menu_odoo_menu.xml [hunk 3]
```diff
@@ -2,10 +2,10 @@
...
         <div class="container">
             <div class="row">
-                <div class="col-md-6 col-lg pt16 pb24">
+                <div class="col-lg pt16 pb24">
```
Comment (qsm-odoo) [inline addons/website/views/snippets/s_mega_menu_odoo_menu.xml:8]: Why changing this? And I am not sure it can be done in a stable commit :grimacing:

File: addons/website/views/snippets/s_mega_menu_thumbnails.xml [hunk 4]
```diff
@@ -54,36 +54,32 @@
...
+                        <i class="fa fa-shopping-basket mr-2"></i> Pickup in store
                     </p>
                 </div>
-                <div class="col-12 col-md-6 col-lg d-flex justify-content-center align-items-center py-2">
+                <div class="col d-flex justify-content-center align-items-center py-2">
                     <p class="text-muted align-middle m-0">
-                        <i class="s_mega_menu_thumbnails_icon fa fa-truck mr-2"></i> Express delivery
+                        <i class="fa fa-truck mr-2"></i> Express delivery
                     </p>
                 </div>
```
Comment (qsm-odoo) [inline addons/website/views/snippets/s_mega_menu_thumbnails.xml:79]: Definitely something wrong here, should not be done in stable

File: addons/web_editor/static/src/js/editor/snippets.editor.js [hunk 5]
```diff
@@ -1568,11 +1568,19 @@ var SnippetsMenu = Widget.extend({
...
-            $selectorChildren = $openModal.find($selectorChildren);
+        // If a modal or a dropdown is open, the drop zones must be created
+        // only in this element.
+        const $editableArea = self.getEditableArea();
+        let $open = $editableArea.find('.modal:visible');
+        if (!$open.length) {
+            $open = $editableArea.filter('.dropdown-menu.show').parent();
+            if (!$open.length) {
+                $open = $editableArea.find('.dropdown-menu.show').parent();
+            }
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js:1579]:
```javascript
$open = $editableArea.find('.dropdown-menu.show').addBack('.dropdown-menu.show').parent();
```
?

File: addons/web_editor/static/src/js/editor/snippets.editor.js [hunk 6]
```diff
@@ -1568,11 +1568,19 @@ var SnippetsMenu = Widget.extend({
...
+        }
+        if ($open.length) {
+            $selectorSiblings = $open.find($selectorSiblings);
+            $selectorChildren = $open.find($selectorChildren);
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.editor.js:1583]: This commit should be done in a separate PR in 13.0 where the bug already occurs

File: addons/website/static/tests/tours/edit_megamenu.js [hunk 7]
```diff
@@ -42,5 +43,29 @@ tour.register('edit_megamenu', {
...
+            this.$anchor[0].dispatchEvent(new InputEvent('input', {bubbles: true, inputType: 'insertParagraph'}));
+            actions.text('New Menu Item', this.$anchor[0]);
+        },
+    },
+    {
+        content: "Menu Item should have been renamed",
+        trigger: '.o_mega_menu h4:contains("New Menu Item")',
+        run: function () {}, // it's a check
+    },
+    ...wTourUtils.clickOnSave(),
```
Comment (qsm-odoo) [inline addons/website/static/tests/tours/edit_megamenu.js:69]: Cool to complete this test but makes me wonder how you tested it... as I don't see it being run by a python test :grimacing: The commit which [introduces it ](https://github.com/odoo/odoo/commit/fef5c6eff78101d3903f3e8f2a40b92dc7dc9d7b) does not start it in any way... maybe check with them? :cry:

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/web_editor/static/src/js/editor/snippets.editor.js (hunk 6 above)
Comment (younn-o) [reply addons/web_editor/static/src/js/editor/snippets.editor.js:1583]: The fix is now targeting 13.0 here: https://github.com/odoo/odoo/pull/80603

File: addons/website/static/tests/tours/edit_megamenu.js (hunk 7 above)
Comment (younn-o) [reply addons/website/static/tests/tours/edit_megamenu.js]: I was playing it in `?debug=tests` from the console with `odoo.startTour('edit_megamenu')` but indeed it was never played from python with the other at_install tests...

File: addons/website/static/src/snippets/s_mega_menu_little_icons/000.scss [hunk 8]
```diff
@@ -6,21 +6,7 @@
         }
     }
 
-    .s_mega_menu_gray_area:last-child {
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_mega_menu_little_icons/000.scss:9]: You lost the `:last-child` part, thus changing the stable style, we can't :wink: Same at the other location

File: addons/website/static/src/snippets/s_mega_menu_odoo_menu/000.scss [hunk 9]
```diff
@@ -2,7 +2,10 @@
 .s_mega_menu_odoo_menu:not([data-vcss]) {
     .s_mega_menu_odoo_menu_footer {
         // Apply color transparency to match with the preset used
-        border-color: rgba(0, 0, 0, .05) !important;
-        background: rgba(0, 0, 0, .05) !important;
+        border-color: rgba(0, 0, 0, .05);
+
+        .row > div {
+            background-color: rgba(0, 0, 0, .05);
+        }
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_mega_menu_odoo_menu/000.scss:9]: This one is also a stable style change... but in this case I judge it's ok as it fixes more problems that the potential "break"

File: addons/website/views/snippets/snippets.xml [hunk 10]
```diff
@@ -462,8 +462,8 @@
     </div>
 
     <!-- Background -->
-    <t t-set="only_bg_color_selector" t-value="'section .row > div, .s_text_highlight'"/>
-    <t t-set="only_bg_color_exclude" t-value="'.s_col_no_bgcolor, .s_col_no_bgcolor.row > div, .s_masonry_block .row > div, .s_color_blocks_2 .row > div, .o_mega_menu .row > div, .s_image_gallery .row > div'"/>
+    <t t-set="only_bg_color_selector" t-value="'section .row > div, .s_text_highlight, .s_mega_menu_thumbnails_footer'"/>
+    <t t-set="only_bg_color_exclude" t-value="'.s_col_no_bgcolor, .s_col_no_bgcolor.row > div, .s_masonry_block .row > div, .s_color_blocks_2 .row > div, .s_image_gallery .row > div, .o_mega_menu .s_mega_menu_gray_area'"/>
```
Comment (qsm-odoo) [inline addons/website/views/snippets/snippets.xml:466]: Is it actually useful to exclude gray areas? I understand the right side of the column would still be colored gray but I don't think it's really a problem. Besides, if the user re-order the column, then it should definitely not be excluded. So I think you can simply remove this.

File: addons/website/static/src/js/content/menu.js [hunk 11]
```diff
@@ -178,8 +178,10 @@ const BaseAnimatedHeader = animations.Animation.extend({
         }
 
         // Close opened menus
-        this.$dropdowns.removeClass('show');
-        this.$navbarCollapses.removeClass('show').attr('aria-expanded', false);
+        if (scroll) {
```
Comment (qsm-odoo) [inline addons/website/static/src/js/content/menu.js:181]: I guess it's an acceptable solution... but you could still edit mega menus with the page scrolled. But if there really is no other way, it's ok, however: is not a comment useful to understand why this is necessary?

File: addons/website/tests/test_ui.py [hunk 12]
```diff
@@ -275,3 +275,6 @@ def test_13_editor_focus_blur_unit_test(self):
 
     def test_14_carousel_snippet_content_removal(self):
         self.start_tour("/", "carousel_content_removal", login='admin')
+
+    def test_15_edit_megamenu(self):
+        self.start_tour("/", "edit_megamenu", login='admin')
```
Comment (qsm-odoo) [inline addons/website/tests/test_ui.py:280]: Did you ping the editor team about that? Just to know if I should notify them :wink:

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/website/tests/test_ui.py (hunk 12 above)
Comment (younn-o) [reply addons/website/tests/test_ui.py:280]: I did not check with the editor team for this commit...
@dmo-odoo @Goaman do you think it is okay to set the editor observer unactive when toggling the mega menu to avoid recording the "show" class addition/deletion ? Maybe it should be the case for all the dropdowns

File: addons/website/static/src/js/content/menu.js (hunk 11 above)
Comment (younn-o) [reply addons/website/static/src/js/content/menu.js:181]: I changed for not closing the mega menu on a scroll, I had not thought about it, thanks !

File: addons/website/static/src/js/content/menu.js [hunk 13]
```diff
@@ -32,7 +32,7 @@ const BaseAnimatedHeader = animations.Animation.extend({
     start: function () {
         this.$main = this.$el.next('main');
         this.isOverlayHeader = !!this.$el.closest('.o_header_overlay, .o_header_overlay_theme').length;
-        this.$dropdowns = this.$el.find('.dropdown, .dropdown-menu');
+        this.$dropdowns = this.$el.find('.dropdown:not(:has(> .o_mega_menu)), .dropdown-menu:not(.o_mega_menu)');
```
Comment (qsm-odoo) [inline addons/website/static/src/js/content/menu.js:35]: @younn-o What I meant is that in edit mode, it is possible to scroll the page **then** opening a mega menu (and thus still having the close-on-enter bug with your previous fix). With this change, you change a functionnal behavior which I am not sure make much sense: why do we close dropdown on scroll but not the mega menus :thinking: I am not sure that logic make sense.

Also I tested and it feels strange: depending on the effect you chose for your header scroll, it is nice or ok or seems buggy. With a disappear effect for example, it makes no sense. The mega menu stays open on scroll but the header disappear leaving a floating menu...

File: addons/website/tests/test_ui.py (hunk 12 above)
Comment (qsm-odoo) [reply addons/website/tests/test_ui.py:280]: @younn-o Good to ask for the observer stuff but I was not commenting those line,s I meant notify them about this test.

@dmo-odoo @Zinston Just FWI you added a JS test [here](https://github.com/odoo/odoo/commit/fef5c6eff78101d3903f3e8f2a40b92dc7dc9d7b) without adding a python test so if I am not mistaken I don't think it was testing anything :wink: We're adding it here :+1:

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
@younn-o Things to review. Also the test addition will conflict with your unrelated link PR  which is also adding a test at the same place, move this back to RD Dev - Green once you adapted and rebased the branch once the other PR has been merged :+1: Thanks :+1:

File: addons/website/tests/test_ui.py (hunk 12 above)
Comment (dmo-odoo) [reply addons/website/tests/test_ui.py:280]: To be honest, I don't remember. I guess it's possible we used to run this test manually and forgot to add the python call to have it tested on runbot. Thanks !

**Review by qsm-odoo — APPROVED** [review]
@younn-o I added this diff to the first commit here: https://github.com/odoo/odoo/compare/b8b5c3d6c0222e1e0e53bd616e732e71cac0c101..f9ad54928afa40b753ed8b276d63e5d5a057d473 ... you added the possibility to change the color via the editor, adapted the css rules too... but it still did not work, I wonder how you did not see it :thinking:  Anyways, there was a series of problem preventing the color update, I hacked my way through it using the same ::before element trick so the fix is stable (arguably, the fix is not entirely stable since the small XML change and existing users won't get this fix but still I think it's ok to merge in stable here since no design change, extremely small risk of breaking the view xpath).

Anyways :)

@robodoo rebase-ff r+
